### PR TITLE
GC: Use upstream patch for green threads support

### PR DESCRIPTION
**Pull request by bcardiff**
This requires also an update of the distribution-scripts (~~reason why this PR is draft~~).
A compiler with the updated GC 8.0.4 + upstream patch will be available in circleci for a smoke testing. So far it seems to work locally.

The preview_mt jobs will fail with compilation issues since the 0.31.0 has the old patch bundled.

As of the changes itself, bdw-gc patch ended returning an opaque thread handler that is stored now in `Thread.gc_thread_handler` and a `StackBase` structure that is the stackbottom in most architectures. In IA-64 includes the reg_base, but we don't support IA-64. Since the fibers only knows about the stackbottom, I preferred to leave the reg_base commented.

Depends on: https://github.com/crystal-lang/distribution-scripts/pull/47

Fixes: #8213

### CODE REVIEW

**Comment by bcardiff** [comment]
A build with the upstream patch is available in `snap refresh crystal --channel=edge/ci-update-gc-dev`.

Unsigned rpm, deb packages and tar.gz can be downloaded from https://circleci.com/gh/crystal-lang/crystal/29489#artifacts/containers/0 in case someone is willing to smoke on their environment.

**Comment by bcardiff** [comment]
On my side the smoke test went well. I've run the benchmarks and nothing explodes, memory consumption was similar than before.
 
If someone else wants to confirm/try and review the changes we can move forward here.

**Comment by bararchy** [comment]
@bcardiff works for us :+1:

File: src/gc/boehm.cr [hunk 1]
```diff
@@ -17,6 +17,13 @@ lib LibGC
   alias SizeT = LibC::SizeT
   alias Word = LibC::ULong
 
+  struct StackBase
+    mem_base : Void*
+    # reg_base : Void* should be used also for IA-64 when/if supported
```
Comment (RX14) [inline src/gc/boehm.cr]: We will never support IA-64, the architecture is thoroughly dead.

**Comment by anatol** [comment]
I tested this patch with upstream GC changes at Arch Linux and it looks good. Thank you for the work.

One the patch lands crystal repo I'll push a new Arch `crystal` package.

**Comment by bararchy** [comment]
@bcardiff seems both @anatol and our team verified it's working, this will really help up streamline development.
Can we cut a new version? :pray:

**Pull request merged**